src/main/java/org/opensearch/neuralsearch/processor/NormalizationProcessor.java

```diff
@@ -0,0 +1,204 @@
+import org.opensearch.search.SearchPhaseResult;
+import org.opensearch.search.internal.SearchContext;
+import org.opensearch.search.pipeline.SearchPhaseResultsProcessor;
+import org.opensearch.search.query.QuerySearchResult;
+
+import com.google.common.annotations.VisibleForTesting;
+
+/**
+ * Processor for score normalization and combination on post query search results. Updates query results with
+ * normalized and combined scores for next phase (typically it's FETCH)
+ */
+@Log4j2
+public class NormalizationProcessor implements SearchPhaseResultsProcessor {
+    public static final String TYPE = "normalization-processor";
+    public static final String NORMALIZATION_CLAUSE = "normalization";
+    public static final String COMBINATION_CLAUSE = "combination";
+    public static final String TECHNIQUE = "technique";
+
+    private final String tag;
+    private final String description;
+    @VisibleForTesting
+    @Getter(AccessLevel.PACKAGE)
+    final ScoreNormalizationTechnique normalizationTechnique;
+    @Getter(AccessLevel.PACKAGE)
+    final ScoreCombinationTechnique combinationTechnique;
+
+    /**
+     * Need all args constructor to validate parameters and fail fast
+     * @param tag
+     * @param description
+     * @param normalizationTechnique
+     * @param combinationTechnique
+     */
+    public NormalizationProcessor(
+        final String tag,
+        final String description,
+        final String normalizationTechnique,
+        final String combinationTechnique
+    ) {
+        this.tag = tag;
+        this.description = description;
+        validateParameters(normalizationTechnique, combinationTechnique);
+        this.normalizationTechnique = ScoreNormalizationTechnique.valueOf(normalizationTechnique);
+        this.combinationTechnique = ScoreCombinationTechnique.valueOf(combinationTechnique);
+    }
+
+    /**
+     * Method abstracts functional aspect of score normalization and score combination. Exact methods for each processing stage
+     * are set as part of class constructor
+     * @param searchPhaseResult {@link SearchPhaseResults} DTO that has query search results. Results will be mutated as part of this method execution
+     * @param searchPhaseContext {@link SearchContext}
+     * @param <Result>
+     */
+    @Override
+    public <Result extends SearchPhaseResult> void process(
+        final SearchPhaseResults<Result> searchPhaseResult,
+        final SearchPhaseContext searchPhaseContext
+    ) {
+        if (searchPhaseResult instanceof QueryPhaseResultConsumer) {
+            QueryPhaseResultConsumer queryPhaseResultConsumer = (QueryPhaseResultConsumer) searchPhaseResult;
+            Optional<SearchPhaseResult> maybeResult = queryPhaseResultConsumer.getAtomicArray()
+                .asList()
+                .stream()
+                .filter(Objects::nonNull)
+                .findFirst();
+            if (isNotHybridQuery(maybeResult)) {
```

Comment:
we should make the condition as rather than making it negation.

`if(hybridQueryResults(maybeResults))`

Reply:
Negation is needed to fail fast. Options are to negate in the if clause (```if (!hybridQueryResults(maybeResults))```) which is hard to read and lass error prone, or encapsulate that in the method.